src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp

```diff
@@ -0,0 +1,227 @@
+/**
+ * @file td_idf_encoding_policy.hpp
+ * @author Jeffin Sam
+ *
+ * Definition of the TfIdfEncodingPolicy class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_ENCODING_POLICIES_TF_IDF_ENCODING_POLICY_HPP
+#define MLPACK_CORE_DATA_ENCODING_POLICIES_TF_IDF_ENCODING_POLICY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/data/string_encoding_policies/policy_traits.hpp>
+#include <mlpack/core/data/string_encoding.hpp>
+namespace mlpack {
+namespace data {
+/**
+ * Definition of the TfIdfEncodingPolicy class.
+ *
+ * Tf means term-frequency while tf-idf means term-frequency times inverse
+ * document-frequency. This is a common term weighting scheme in information
+ * retrieval, that has also found good use in document classification.
+ * The goal of using tf-idf instead of the raw frequencies of occurrence of a
+ * token in a given document is to scale down the impact of tokens that occur
+ * very frequently in a given corpus and that are hence empirically less
+ * informative than features that occur in a small fraction of the training
+ * corpus.
+ * TfIdfEncodingPolicy is used as a helper class for StringEncoding.
+ * The encoder assigns a tf-idf number to each unique token and treat 
+ * the dataset as categorical. The tokens are labeled in the order of their
+ * occurrence in the input dataset.
+ */
+class TfIdfEncodingPolicy
+{
+ public:
+  /* 
+  * Enum Class used to identify the type of tf encoding
+  *
+  * Follwing are the defination of the types
+  * binary : binary weighting scheme (0,1)
+  * rawCount : raw count weighting scheme (count of token for every row)
+  * termFrequency : term frequency weighting scheme (count / length(row))
+  * subinerTf : logarthimic weighting scheme (log(tf) + 1) 
+  * 
+  */
+  enum tfTypes
+  {
+    rawCount,
+    binary,
+    sublinearTf,
+    termFrequency,
+  };
```

Comment:
It's not mentioned in the mlpack style guide. But we use the following style in the rest of the code:
```
enum TfTypes
{
  RAW_COUNT,
  BINARY,
  SUBLINEAR_TF,
  TERM_FREQUENCY,
};
```

Reply:
Probably it's better to use `enum class` instead of `enum` since `enum class` can't be initialized from an integer without `static_cast`.